File: server/routes/command.js
import express from 'express';
import { ensureFullyApproved } from '../middleware/auth.js';
import { lockService } from '../services/LockService.js';

const router = express.Router();

/**
 * POST /command - Single endpoint for all lock commands
 * Keeps sensitive API keys hidden from frontend
 */
router.post('/', ensureFullyApproved, async (req, res) => {
  const { action, bikeId, sessionId } = req.body;
  const userId = req.user.id;

  try {
    switch (action) {
      case 'unlock_chain': {
        // Step 1: Unlock TetherSense chain
        if (!bikeId) {
          return res.status(400).json({ error: 'bikeId is required' });
        }

        // Check if user already has an active ride
        const existingSession = lockService.getActiveSession(userId);
        if (existingSession) {
          return res.status(400).json({
            error: 'You already have an active ride',
            sessionId: existingSession.sessionId
          });
        }

        const chainResult = await lockService.unlockChain(bikeId, userId);
        res.json(chainResult);
        break;
      }

      case 'unlock_wheel': {
        // Step 2: Unlock rear wheel after chain confirmation
        if (!sessionId) {
          return res.status(400).json({ error: 'sessionId is required' });
        }

        const wheelResult = await lockService.unlockWheel(sessionId);
        res.json(wheelResult);
        break;
      }

      case 'lock': {
        // End ride - lock the bike
        if (!sessionId) {
          return res.status(400).json({ error: 'sessionId is required' });
        }

        const lockResult = await lockService.lockBike(sessionId);
        res.json(lockResult);
        break;
      }

      case 'status': {
        // Get current lock status
        if (!bikeId) {
          return res.status(400).json({ error: 'bikeId is required' });
        }

        const status = await lockService.getLockStatus(bikeId);
        res.json(status);
        break;
      }

      case 'active_session': {
        // Get user's active session
        const session = lockService.getActiveSession(userId);
        res.json({ session });
        break;
      }

      default:
        res.status(400).json({ error: 'Invalid action' });
    }
  } catch (error) {
    console.error('Command error:', error);
    res.status(500).json({ error: error.message || 'Command failed' });
  }
});

// Webhook endpoint for LINKA auto-lock events
router.post('/webhook', async (req, res) => {
  // Verify webhook signature in production
  const signature = req.headers['x-linka-signature'];

  // For prototype, accept all webhooks
  try {
    const result = await lockService.handleWebhook(req.body);
    res.json(result);
  } catch (error) {
    console.error('Webhook error:', error);
    res.status(500).json({ error: 'Webhook processing failed' });
  }
});

export default router;

File: api/_routes/command.py
from datetime import datetime

from fastapi import APIRouter, HTTPException, Request

from api._linka import call_linka
from api.services.ride_service import ride_service

router = APIRouter()


@router.post("/api/command")
async def command(request: Request):
    body       = await request.json()
    action     = body.get("action")
    bike_id    = body.get("bikeId")
    session_id = body.get("sessionId")

    if action == "open":
        if not bike_id:
            raise HTTPException(status_code=400, detail="bikeId is required")
        return await ride_service.start_ride(bike_id)

    if action == "unlock_chain":
        if not bike_id:
            raise HTTPException(status_code=400, detail="bikeId is required")
        return await ride_service.unlock_chain(bike_id)

    if action == "unlock_wheel":
        if not session_id:
            raise HTTPException(status_code=400, detail="sessionId is required")
        return await ride_service.unlock_wheel(session_id)

    if action == "lock":
        if not session_id:
            raise HTTPException(status_code=400, detail="sessionId is required")
        return await ride_service.lock_ride(session_id)

    if action == "status":
        if not bike_id:
            raise HTTPException(status_code=400, detail="bikeId is required")
        try:
            return await call_linka(f"/device_status/{bike_id}")
        except Exception:
            return {"bikeId": bike_id, "chainLocked": True, "wheelLocked": True,
                    "batteryLevel": 85, "lastUpdated": datetime.utcnow().isoformat()}

    raise HTTPException(status_code=400, detail="Invalid action")


@router.post("/api/command/webhook")
async def command_webhook(request: Request):
    payload = await request.json()
    bike_id = payload.get("bikeId")
    event   = payload.get("event")

    if event == "chain_locked" and bike_id:
        result = await ride_service.handle_webhook_lock(bike_id)
        if result:
            return result

    return {"success": True, "message": "Webhook processed"}
